src/relay/transforms/merge_compiler_regions.cc

```diff
@@ -0,0 +1,352 @@
+
+  Expr Annotate(const Expr& expr) {
+    // Its a function that is being passed on to annotate
+    func_ = Downcast<Function>(expr);
+
+    // Corner Case CC1 : If the last node does not belong
+    // to a region nede to add a compiler_end
+    auto region = regions_->GetRegion(func_->body);
+    auto mutated_expr = this->VisitExpr(expr);
+    if (!region.defined()) {
+      func_ = Downcast<Function>(mutated_expr);
+      // CC1 : add that compiler end after mutation
+      auto body = AddCompilerEnd_(func_->body);
+      func_ = Function(func_->params, body,
+                       body->checked_type_, {}, DictAttrs());
+      return Downcast<Expr>(func_);
+    } else {
+      return mutated_expr;
+    }
+  }
+
+  /*! \brief This function adds compiler ends to nodes that
+   * have a region AND they should not be arguments of the
+   * original function
+   * \param expr
+   * \return expr
+   */
+  Expr AddCompilerEnd(const Expr& expr) {
+    auto region = regions_->GetRegion(expr);
+    auto visited_expr = VisitExpr(expr);
+
+    // The compiler ends are added to nodes that does have a region
+    // AND they should not be arguments of the original function
+    if (!region.defined() &&
+       std::find(func_->params.begin(),
+                 func_->params.end(), visited_expr)
+                   == func_->params.end()) {
+      return AddCompilerEnd_(visited_expr);
+    } else {
+      return visited_expr;
+    }
+  }
+
+  Expr AddCompilerEnd_(const Expr& expr) {
+    const auto* end_op =
+      runtime::Registry::Get("relay.op.annotation._make.compiler_end");
+    CHECK(end_op);
+    Expr end = (*end_op)(expr, target_);
+    return end;
+  }
+
+  Expr VisitExpr_(const CallNode* call) final {
+    auto op_node = call->op.as<OpNode>();
+    auto ret = GetRef<Call>(call);
+
+    Array<Expr> args;
+
+    // Add compiler ends if the parent is supported
+    for (auto arg : call->args) {
+      args.push_back(AddCompilerEnd(arg));
+    }
+
+    if (op_node == nullptr || call->attrs.as<CompilerAttrs>() == nullptr) {
+      // Skip annotatation ops, only add default compiler to actual compute nodes
+
+      auto region = regions_->GetRegion(ret);
+      if (!region.defined()) {
+        // if the current node does not belong to annotated region
+        // annotate the all incoming edges (args)
+        // with "default" compile_begin and compiler_end annotations.
+        tvm::Array<tvm::relay::Expr> compiler_begins;
+        for (auto arg : args) {
+          const auto* begin_op =
+            runtime::Registry::Get("relay.op.annotation._make.compiler_begin");
+          CHECK(begin_op);
+          Expr begin = (*begin_op)(arg, target_);
+          compiler_begins.push_back(begin);
+        }
+        Expr update_call = Call(call->op, compiler_begins, call->attrs);
+        return update_call;
+      } else {
+        return Call(call->op, args, call->attrs);
+      }
+    } else {
+      return Call(call->op, args, call->attrs);
+    }
```

Comment:
Likely, we don't need the last two `else`, but just have `return Call(call->op, args, call->attrs);` in the end